src/RollerPeriphery.sol

```diff
@@ -22,152 +25,182 @@ contract RollerPeriphery is Trust {
 
-        ERC20(adapter.underlying()).safeTransfer(receiver, underlyingOut);
+        _transferUnderlying(roller, receiver);
     }
 
-    /// @notice Withdraw asset from vault with slippage protection
-    /// @param roller AutoRoller vault
-    /// @param assets Amount of asset requested for withdrawal
-    /// @param receiver Destination address for the returned assets
-    /// @param maxSharesOut Maximum amount of shares burned
-    /// @return shares Number of shares to redeem
-    function withdraw(AutoRoller roller, uint256 assets, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
-        if ((shares = roller.withdraw(assets, receiver, msg.sender)) > maxSharesOut) {
-            revert MaxSharesError();
-        }
-    }
+    
 
-    /// @notice Withdraw asset from vault and convert to underlying, with slippage protection
+    /// @notice Withdraw asset from vault and convert to token, with slippage protection
+    /// @dev If withdrawing to other token than underlying or asset, amtOut should be the expected amount resulting from the conversion from to the token to underlying 
     /// @param roller AutoRoller vault
-    /// @param underlyingOut Amount of underlying requested for withdrawal
+    /// @param amtOut Amount of tokens requested for withdrawal (in underlying if withdrawing any token or underlying)
     /// @param receiver Destination address for the returned underlying
-    /// @param maxSharesOut Maximum amount of shared burned
-    /// @return shares Number of shares to redeem
-    function withdrawUnderlying(AutoRoller roller, uint256 underlyingOut, address receiver, uint256 maxSharesOut) external returns (uint256 shares) {
+    /// @param maxSharesOut Maximum amount of shares burned
+    /// @param permit Permit message to pull shares from caller
+    /// @param quote Swap quote for converting underlying to token
+    /// @return shares Number of shares to redeem,
+    function withdraw(AutoRoller roller, uint256 amtOut, address receiver, uint256 maxSharesOut, PermitData calldata permit, SwapQuote calldata quote) external returns (uint256 shares) {
         AdapterLike adapter = AdapterLike(address(roller.adapter()));
+        uint256 assetOut = amtOut;
+
+        if (address(quote.buyToken) == address(adapter.underlying())) {
+            // asset converted from underlying (round down)
+            assetOut = amtOut.divWadDown(adapter.scale());
+        } else if (address(quote.buyToken) != address(adapter.target())) {
+            // TODO: here we would need to convert the amount of token to asset (target)
+            // 2 options:
+            // a) amtOut would be the amount of tokens requested for withdrawal and 
+            // we send the price of underlying-token as a param (maybe inside the quote) 
+            // so we can calculate token-tounderlying-to-asset
+            // b) amtOut is the price of token-underlying so then we can calculate the -to-asset by dividing for scale
```

Comment:
This would mean we always have the user pass in amtOut in underlying terms (ie they figure out underlying -> token conversion price themselves off-chain)? Think i'm supportive of that if so. Do you lean towards one or the other?